### PR TITLE
Remove PolymorphicFunctionType

### PR DESCRIPTION
**Pull request by slavapestov**
For those following along at home, this is a continuation of the work in https://github.com/apple/swift/pull/4849 and earlier commits like https://github.com/apple/swift/commit/da4af641a7581764ed1e0365765e2d57da60b9ab. PolymorphicFunctionType is the contextual type equivalent of GenericFunctionType, a polymorphic function signature but written in terms of archetypes. It didn't make much sense in the first place, and GenericFunctionType had a FIXME comment explaining it should replace PolymorphicFunctionType -- since 2013 :-)

There are three big pieces here:

- Ensuring all ValueDecls have an interface type assigned, instead of getInterfaceType() punting to the contextual type if an interface type is not available
- Ensuring getType() is not called on polymorphic declarations at all, with getInterfaceType() used everywhere instead
- Removing the code that sets the contextual type on polymorphic declarations, which means that PolymorphicFunctionType is now dead and fully subsumed by GenericFunctionType

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: validation-test/IDE/crashers_fixed/087-swift-declcontext-getparentmodule.swift [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// RUN: %target-swift-ide-test -code-completion -code-completion-token=A -source-filename=%s
+// REQUIRES: asserts
```
Comment (DougGregor) [inline validation-test/IDE/crashers_fixed/087-swift-declcontext-getparentmodule.swift]: doesn't require asserts and more

**Review by DougGregor — COMMENTED** [review]
I agree that the uses of `SubstitutedType` are fragile. In the first case, `getMemberSubstitutions()` seems like a far simpler and more correct solution.

For accessibility and availability checking, maintaining proper sugar in the `Type` (with all of the appropriate declaration references as written) would make it far easier. Or, better yet, use the `TypeRepr` for these checks: then we have all of the source-location information.

**Review by DougGregor — COMMENTED** [review]
Kinda surprised we can do this, because IIRC we haven't done anything to preclude this recursion elsewhere.

**Comment by DougGregor** [comment]
I browsed through everything thus far, and it's looking great. I hadn't realized we were relying on `SubstitutedType` for semantics in some places, which is... a bit disturbing. Definitely some cleanup to be done there.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Comment by slavapestov** [comment]
@swift-ci Please test macOS

**Comment by slavapestov** [comment]
@swift-ci Please test macOS

**Comment by slavapestov** [comment]
@DougGregor This is now at the point where check-swift-validation passes on my Mac, however it looks like I broke something in SwiftPM, and lldb needs changes as usual. Progress :)

**Comment by slavapestov** [comment]
@swift-ci Please test macOS

**Comment by slavapestov** [comment]
@swift-ci Please test macOS

**Comment by slavapestov** [comment]
@swift-ci Please smoke test macOS

**Comment by slavapestov** [comment]
@swift-ci Please smoke test macOS

**Comment by slavapestov** [comment]
@swift-ci Please smoke test macOS

**Comment by slavapestov** [comment]
@swift-ci Please smoke test macOS

**Comment by slavapestov** [comment]
@swift-ci Please smoke test macOS

**Comment by slavapestov** [comment]
@swift-ci Please smoke test macOS

**Comment by slavapestov** [comment]
@swift-ci Please smoke test macOS

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Comment by slavapestov** [comment]
@swift-ci Please smoke test macOS

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Comment by slavapestov** [comment]
@swift-ci Please smoke test macOS

**Comment by slavapestov** [comment]
@swift-ci Please test Linux

**Comment by slavapestov** [comment]
@swift-ci Please test macOS

**Comment by DougGregor** [comment]
Waahoooooooo!

**Pull request merged**